Modules/grpmodule.c

```diff
@@ -119,16 +120,47 @@ grp_getgrgid_impl(PyObject *module, PyObject *id)
         }
         Py_DECREF(py_int_id);
     }
+#ifdef HAVE_GETGRGID_R
+    Py_BEGIN_ALLOW_THREADS
+    int status, bufsize = NSS_BUFLEN_GROUP;
+    struct group grp;
 
-    if ((p = getgrgid(gid)) == NULL) {
+    buf = PyMem_RawMalloc(bufsize);
+
+    do {
+        status = getgrgid_r(gid, &grp, buf, bufsize, &p);
+        if (p != NULL || status != ERANGE) {
+            break;
+        }
+        if ((bufsize << 1) > (NSS_BUFLEN_GROUP << 10)) {
+            status = ERANGE;
+            break;
+        }
+        bufsize <<= 1;
+        buf = PyMem_RawRealloc(buf, bufsize);
+    } while (1);
+
+    if (status != 0 || p == NULL) {
+        PyMem_RawFree(buf);
+        p = NULL;
+    }
+    Py_END_ALLOW_THREADS
+#else
+    p = getgrgid(gid);
+#endif
+    if (p == NULL) {
         PyObject *gid_obj = _PyLong_FromGid(gid);
         if (gid_obj == NULL)
             return NULL;
```

Comment:
You leak buf memory here.

Reply:
I am failing to see how. If p == NULL it will free buf in the if statement before.

Reply:
Oh. I was confused because one RawFree in done when the GIL is released, and the other one is done when we hold the GIL. I would suggest to release buf when we hold the GIL, and change the previous if as:

```
    if (status != 0) {
        p = NULL;
    }
```